pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+	endpoints := []*feddnsv1a1.Endpoint{}
+
+	commonPrefix := strings.Join([]string{dnsObject.Name, dnsObject.Namespace, federation, "svc"}, ".")
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		zone := clusterDNS.Zone
+		region := clusterDNS.Region
+
+		dnsNames := []string{
+			strings.Join([]string{commonPrefix, zone, region, dnsSuffix}, "."), // zone level
+			strings.Join([]string{commonPrefix, region, dnsSuffix}, "."),       // region level, one up from zone level
+			strings.Join([]string{commonPrefix, dnsSuffix}, "."),               // global level, one up from region level
+			"", // nowhere to go up from global level
+		}
+
+		zoneTargets, regionTargets, globalTargets := getHealthyTargets(zone, region, dnsObject)
+		targets := [][]string{zoneTargets, regionTargets, globalTargets}
+
+		for i, target := range targets {
+			endpoint, err := generateEndpoint(dnsNames[i], target, dnsNames[i+1])
+			if err != nil {
+				return nil, err
+			}
+			endpoints = append(endpoints, endpoint)
+		}
+	}
+
+	return endpoints, nil
+}
+
+// getHealthyTargets returns the hostnames and/or IP addresses of healthy endpoints for the service, at a zone, region and global level (or an error)
+func getHealthyTargets(zone, region string, dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord) (zoneTargets, regionTargets, globalTargets feddnsv1a1.Targets) {
+	// If federated dnsObject is deleted, return empty endpoints, so that DNS records are removed
+	if dnsObject.DeletionTimestamp != nil {
+		return zoneTargets, regionTargets, globalTargets
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		if clusterDNS.Zone == zone {
+			zoneTargets = append(zoneTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		}
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		if clusterDNS.Region == region {
+			regionTargets = append(regionTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		}
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		globalTargets = append(globalTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+	}
+
+	return zoneTargets, regionTargets, globalTargets
+}
+
+func generateEndpoint(name string, targets feddnsv1a1.Targets, uplevelCname string) (ep *feddnsv1a1.Endpoint, err error) {
+	ep = &feddnsv1a1.Endpoint{
+		DNSName:   name,
+		RecordTTL: minDNSTTL,
+	}
+
+	if len(targets) > 0 {
+		targets, err = getResolvedTargets(targets, netWrapper)
+		if err != nil {
+			return nil, err
+		}
+		ep.Targets = targets
+		ep.RecordType = RecordTypeA
+	} else {
+		ep.Targets = []string{uplevelCname}
+		ep.RecordType = RecordTypeCNAME
+	}
+
+	return ep, nil
+}
+
+// getResolvedTargets performs DNS resolution on the provided slice of endpoints (which might be DNS names
+// or IPv4 addresses) and returns a list of IPv4 addresses.  If any of the endpoints are neither valid IPv4
+// addresses nor resolvable DNS names, non-nil error is also returned (possibly along with a partially
```

Comment:
(No action required) Why is a partially complete list of resolved endpoints returned if the result is thrown away on error by the only caller?